File: wrappers/java/src/main/java/com/turbotoken/NativeBridge.java
package com.turbotoken;

/**
 * JNI native method declarations matching jni/turbotoken_jni.c.
 * Package-private -- not part of the public API.
 */
final class NativeBridge {

    static {
        NativeLoader.load();
    }

    private NativeBridge() {}

    static native String version();

    static native void clearRankTableCache();

    static native int[] encodeBpe(byte[] rankBytes, byte[] textBytes);

    static native byte[] decodeBpe(byte[] rankBytes, int[] tokens);

    static native long countBpe(byte[] rankBytes, byte[] textBytes);

    static native long isWithinTokenLimit(byte[] rankBytes, byte[] textBytes, long tokenLimit);

    static native int[] encodeBpeFile(byte[] rankBytes, String filePath);

    static native long countBpeFile(byte[] rankBytes, String filePath);

    static native int[] trainBpeFromChunkCounts(
        byte[] chunks, int[] chunkOffsets, int[] chunkCounts,
        int vocabSize, int minFrequency
    );
}

File: wrappers/kotlin/src/main/kotlin/com/turbotoken/NativeBridge.kt
package com.turbotoken

/**
 * Kotlin-side wrapper around the Java NativeBridge.
 * Delegates all native calls to the Java JNI bridge.
 */
internal object KNativeBridge {

    fun version(): String = NativeBridge.version()

    fun clearRankTableCache() = NativeBridge.clearRankTableCache()

    fun encodeBpe(rankBytes: ByteArray, textBytes: ByteArray): IntArray =
        NativeBridge.encodeBpe(rankBytes, textBytes)

    fun decodeBpe(rankBytes: ByteArray, tokens: IntArray): ByteArray =
        NativeBridge.decodeBpe(rankBytes, tokens)

    fun countBpe(rankBytes: ByteArray, textBytes: ByteArray): Long =
        NativeBridge.countBpe(rankBytes, textBytes)

    fun isWithinTokenLimit(rankBytes: ByteArray, textBytes: ByteArray, tokenLimit: Long): Long =
        NativeBridge.isWithinTokenLimit(rankBytes, textBytes, tokenLimit)

    fun encodeBpeFile(rankBytes: ByteArray, filePath: String): IntArray =
        NativeBridge.encodeBpeFile(rankBytes, filePath)

    fun countBpeFile(rankBytes: ByteArray, filePath: String): Long =
        NativeBridge.countBpeFile(rankBytes, filePath)

    fun trainBpeFromChunkCounts(
        chunks: ByteArray,
        chunkOffsets: IntArray,
        chunkCounts: IntArray,
        vocabSize: Int,
        minFrequency: Int
    ): IntArray = NativeBridge.trainBpeFromChunkCounts(
        chunks, chunkOffsets, chunkCounts, vocabSize, minFrequency
    )
}
